src/backend/distributed/master/master_repair_shards.c

```diff
@@ -37,118 +39,192 @@
 	Oid distributedTableId = shardInterval->relationId;
+	List *colocatedTableList = ColocatedTableList(distributedTableId);
+	ListCell *colocatedTableCell = NULL;
+	List *colocatedShardList = ColocatedShardIntervalList(shardInterval);
+	ListCell *colocatedShardCell = NULL;
 
-	char *relationOwner = NULL;
-	List *shardPlacementList = NIL;
-	ShardPlacement *sourcePlacement = NULL;
-	ShardPlacement *targetPlacement = NULL;
-	WorkerNode *targetNode = NULL;
-	List *ddlCommandList = NIL;
-	bool dataCopied = false;
-	char relationKind = '\0';
+	/* check that user has owner rights in all co-located tables */
+	foreach(colocatedTableCell, colocatedTableList)
+	{
+		Oid colocatedTableId = lfirst_oid(colocatedTableCell);
+		char relationKind = '\0';
+		EnsureTableOwner(colocatedTableId);
 
-	EnsureTableOwner(distributedTableId);
+		relationKind = get_rel_relkind(colocatedTableId);
+		if (relationKind == RELKIND_FOREIGN_TABLE)
+		{
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("cannot repair shard"),
+							errdetail("Repairing shards backed by foreign tables is "
+									  "not supported.")));
+		}
+	}
 
-	/*
-	 * By taking an exclusive lock on the shard, we both stop all modifications
-	 * (INSERT, UPDATE, or DELETE) and prevent concurrent repair operations from
-	 * being able to operate on this shard.
-	 */
-	LockShardResource(shardId, ExclusiveLock);
+	/* we sort colocatedShardList so that lock operations will not cause any deadlocks */
+	colocatedShardList = SortList(colocatedShardList, CompareShardIntervalsById);
 
 	/*
-	 * We've stopped data modifications of this shard, but we plan to move
-	 * a placement to the healthy state, so we need to grab a shard metadata
-	 * lock (in exclusive mode) as well.
+	 * FIXME
+	 * To reviewer, I am not sure about locking structure here;
+	 * Are these correct lock functions?
+	 * Should we take lock on relationId instead of shardId?
+	 * Should we take lock on colocationId?
 	 */
-	LockShardDistributionMetadata(shardId, ExclusiveLock);
-
-	relationOwner = TableOwner(distributedTableId);
-
-	shardPlacementList = ShardPlacementList(shardId);
-	sourcePlacement = SearchShardPlacementInList(shardPlacementList, sourceNodeName,
-												 sourceNodePort);
-	if (sourcePlacement->shardState != FILE_FINALIZED)
+	foreach(colocatedShardCell, colocatedShardList)
 	{
-		ereport(ERROR, (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
-						errmsg("source placement must be in finalized state")));
+		ShardInterval *colocatedShard = (ShardInterval *) lfirst(colocatedShardCell);
+		uint64 colocatedShardId = colocatedShard->shardId;
+
+		/*
+		 * We've stopped data modifications of this shard, but we plan to move
+		 * a placement to the healthy state, so we need to grab a shard metadata
+		 * lock (in exclusive mode) as well.
+		 */
+		LockShardDistributionMetadata(colocatedShardId, ExclusiveLock);
+
+		EnsureShardCanBeRepaired(colocatedShardId, sourceNodeName, sourceNodePort,
```

Comment:
Check for unrestrictedCopy is done inside EnsureShardCanBeRepaired. In current design we do not need EnsureShardCanBeCopeid. However, as you mentioned in other comment as well if we want to remove that logic from EnsureShardCanBeCopeid, we sould add EnsureShardCanBeCopeid and call one of them based on value of unrestrictedCopy.

I think among these two designs, (making check in EnsureShardCanBeCopeid or here), yours would be more clear and easy to understand. Therefore I will continue with your approach.